### DISCUSSION TITLE
Emulating button presses

### DISCUSSION THREAD

**Discussion by Festivejelly** [kachurovskiy/nanoels#154] [Ideas]
I've got a bit of time of the Christmas holidays to look at implementing a way of emulating the button presses for the controller that we discussed a short while ago.

Im building the GUI and the API in a separate repo so as not to muddy the water here.

But it does require some changes to the code here.

My first thought is to have a function that will be listening for a command sent to it.

1.  handleSerialCommands()

```
void handleSerialCommands() {
  if (Serial.available()) {
    String commandLine = Serial.readStringUntil('\n');
    commandLine.trim(); // remove any leading/trailing whitespace

    // split the command line into command and parameters
    int colonIndex = commandLine.indexOf(':');
    String command = commandLine.substring(0, colonIndex);
    String parameters = commandLine.substring(colonIndex + 1);

    if (command == "CMD") {
      // split the parameters into individual key-value pairs
      char* parameterPairs = strtok(parameters.c_str(), ",");
      while (parameterPairs != NULL) {
        // split the key-value pair into key and value
        char* key = strtok(parameterPairs, ":");
        char* value = strtok(NULL, ":");

        if (strcmp(key, "Turning") == 0) {
          // call your Turning function with the appropriate parameters
          Turning(value);
        } else if (strcmp(key, "Facing") == 0) {
          // call your Facing function with the appropriate parameters
          Facing(value);
        } else if (strcmp(key, "Parting") == 0) {
          // call your Parting function with the appropriate parameters
          Parting(value);
        } else if (strcmp(key, "Threading") == 0) {
          // call your Threading function with the appropriate parameters
          Threading(value);
        }

        parameterPairs = strtok(NULL, ",");
      }
    }
  }
}
```

Turning function example:

```
void Turning(char* parameters) {
  // split the parameters into individual key-value pairs
  char* parameterPairs = strtok(parameters, ",");
  while (parameterPairs != NULL) {
    // split the key-value pair into key and value
    char* key = strtok(parameterPairs, ":");
    char* value = strtok(NULL, ":");

    if (strcmp(key, "Passes") == 0) {
      // handle the Passes parameter
      int passes = atoi(value);
      // ... do something with passes ...
    } else if (strcmp(key, "Pitch") == 0) {
      // handle the Pitch parameter
      float pitch = atof(value);
      // ... do something with pitch ...
    } else if (strcmp(key, "Diameter") == 0) {
      // handle the Diameter parameter
      float diameter = atof(value);
      // ... do something with diameter ...
    } else if (strcmp(key, "Length") == 0) {
      // handle the Length parameter
      float length = atof(value);
      // ... do something with length ...
    }

    parameterPairs = strtok(NULL, ",");
  }
}
```

Or I could just use the key events directly by creating an emulating function that essentially does the same thing but doesnt read the keypad. Instead I pass in the parameters.

```
void Turning(char* parameters) {
  // emulate the Turning button press
  processEmulatedKeypadEvent(B_MODE_TURN);

  // emulate the rest of the button presses for the turning operation
  // replace B_PLUS, B_MINUS, etc. with the actual event codes
  processEmulatedKeypadEvent(B_PLUS);
  processEmulatedKeypadEvent(B_MINUS);
  // ... more button presses ...

  // parse the Diameter and Length parameters from the serial command
  // and use them to set the appropriate variables or perform the appropriate actions
  // ... code to parse parameters ...
}

```

this function will read the command from the serial and then run the relevant function to perform the button presses.

For example for turning the command would be: "CMD:Turning,Passes:3,Pitch:1,Diameter:10,Legnth:15"

So the above function would read the value of CMD and call the Turning function

the turning function would take the parameters passed in and execute a series of button presses in the same fashion as we would press them on the controller.

The benefit of doing it this way is that in the GUI we could have a common set of macros or regular operations that we do. Which means we could do 1 press operations.

I know that the lathecode may make some of this redundant but not all people might be comfortable writing code even if it is really basic.

Would just like to gather your thoughts on this plan and if you have any ideas or improvements you would make?

**Comment by kachurovskiy** [comment]
Hi Nicholas, I see you made good use of some generative AI there :)

We should minimize the changes required in `h4.ino` for this effort so that you can progress independently and to reduce impact on those not using this feature. I would be open to always listening on Serial in `taskGcode`, accepting integers via say `=123\n` command and adding them as `int event` to the array that `processKeypadEvent()` will take it out from and process as key press/release.

This way you can keep all of the UI logic on your side (you can change it independently), we don't have to re-implement every mode as an API and you get access to all features available from UI and can even activate/deactivate GCode mode alongside other modes.

The downside is that when button logic changes in future H4 versions, your GUI might start issuing button presses that do unexpected things. This can be safeguarded against by providing the software version in the status response in https://github.com/kachurovskiy/nanoels/blob/main/h4/h4.ino#L1381 that you could receive by sending `?` on the Serial.

I don't have the capacity to prototype this at the moment, yielding to you to make the demo.

**Reply by Festivejelly** [reply]
Haha yes ive been using github copilot in VScode... its only semi useful. It doesnt seem to quite understand the whole context of the problem. But either way i'll do some experimenting. I'll have a look again based on your suggestion, but I agree it would be good to keep as much logic as possible out of the controller code and let the node.js api handle all that and send simple serial commands to the controller.

I'll keep plugging away and see what I come up with. After i get something work ill upload it to a git repo and link it.

**Reply by Festivejelly** [reply]
Making good progress on this so far. Implemented the version check to avoid any clash.

The API will call that on connecting and will disconnect if the version is not supported.

I'm more familiar with TypeScript so do I actually prefer to implement the logic there and just get it to send the key codes it needs to :)
